physical/postgresql: Add support for high availability

louis-paul opened the pull request:
Are there any e2e tests to run for HA? Thanks!

deverton commented:
Would this be better implemented with advisory locks? https://www.postgresql.org/docs/9.1/static/functions-admin.html#FUNCTIONS-ADVISORY-LOCKS

louis-paul commented:
Hi Dan, I initially started working with advisory locks, which would have been better. Unfortunately advisory locks are per-connection, and Go’s SQL package does pooling over multiple connections by default. This makes lock-related calls not produce the same results depending on the connection they use. 

From my point of view, there are only two ways to use advisory locks for Vault HA with PostgreSQL:

 - Restrict Vault to at most a single connection per instance. This is likely to impact performance in larger deployments in a very significant way (have not measured though). 

 - Dedicate a per-instance connection for locking. Using a full connection that is almost always idle is rather wasteful and may put strain on smaller PostgreSQL servers. 

To the best of my knowledge, this pull request has the same characteristics as an advisory lock-based implementation. Let me know if you have ideas on how to improve the situation. 

-- 
Louis-Paul

> On Jun 29, 2017, at 3:33 AM, Dan Everton <notifications@github.com> wrote:
> 
> Would this be better implemented with advisory locks? https://www.postgresql.org/docs/9.1/static/functions-admin.html#FUNCTIONS-ADVISORY-LOCKS
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jefferai commented:
Honestly, it sounds like advisory locks are the correct way to go. Compared to the traffic coming from Vault generally, an additional single connection with low traffic doesn't seem like a bad tradeoff for getting HA capabilities. That might just be my postgres naivete coming through though.

louis-paul commented:
A particularity of PostgreSQL is that client connections are very expensive for servers, more so than for other databases. From my (limited) understanding of the internals of Postgres, each connection is a separate process with a lot of overhead. Connection pooling middleware (pgbouncer or pgpool) is commonly used to have fewer connections with more activity. The PostgreSQL wiki has [a page on connection count tuning](https://wiki.postgresql.org/wiki/Number_Of_Database_Connections).

For reference, Vault seems to currently use 2 connections per instance ([the default from `database/sql`](https://golang.org/src/database/sql/sql.go?s=17997:18026#L652)).

I wanted to highlight the potential tradeoffs; I would be happy to rewrite the pull request using actual locks if you think that's best. Having the opinion of someone with more experience administering PostgreSQL could be valuable as well.

jefferai commented:
Yeah -- I don't have that experience and I don't know that anyone on the Vault team does. I just know that whenever postgres HA has been discussed in the past advisory locks were deemed the correct approach. I have someone in mind to loop in, will try to get them to comment.

sean- commented on website/source/docs/configuration/storage/postgresql.html.md:
```diff
@@ -82,6 +89,9 @@ LANGUAGE plpgsql;
 
+- `lock_table` `(string: "vault_lock")` – Specifies the name of the table to use
+  for high availability locks. Like `table`, this table must already exist.
+
```
Other doc items to add:

```md
* `lock_ttl` (string: "10s") - Specify the duration of the each lease renewal of the leader.  Once the leader has acquired the lock, it will fresh its lock halfway through its lease.  The first time the leader fails to renew its lease it will attempt every `poll_interval` until either successful or it looses its lease.
* `poll_interval` (string: "1s") - Specify how often clients should poll for the lock.
* `lock_schema` (string: "") - Specify the schema name to use.  If specified the `lock_schema` is used to provide the fully-qualified name of the `lock_table` and bypass looking for the table using the connecition's [`search_path`](https://www.postgresql.org/docs/current/static/ddl-schemas.html#DDL-SCHEMAS-PATH).
```

sean- commented on physical/postgresql.go:
```diff
@@ -175,3 +199,143 @@ func (m *PostgreSQLBackend) List(prefix string) ([]string, error) {
+		if err != nil {
+			return nil, err
+		}
+		lockTTL := strconv.Itoa(PostgreSQLLockTTL)
```
Remove `lockTTL`.

Add a small helper function, `relationName()`:

```go
func (m *PostgreSQLLock) relationName() string
  relationName := pq.QuoteIdentifier(m.lockTableName)
  if schemaName == "" {
    relationName = fmt.Sprintf("%s.%s", pq.QuoteIdentifier(m.lockSchemaName), pq.QuoteIdentifier(m.lockTableName))
  }
  return relationName
}
```

Both `lockTableName` and `lockSchemaName` were already `pq.QuoteIdentifier()`'ed, but if this gets moved to a helper function then we can remove the `pq.QuoteIdentifer()` at config time.  This is defensive in advance of a HUP signal handler.

sean- replied:
```go
lockSQL := fmt.Sprintf(`INSERT INTO %s (key, value, vault_id, expiration)
    VALUES ($1, $2, $3, now() + $4::INTERVAL)`, relationName)
```

Instead of concatenating strings, use `fmt.Sprintf()` to construct the SQL.

That way we're no longer constructing SQL by hand.  It's low risk here, but worth observing best practices.

sean- commented on physical/postgresql.go:
```diff
@@ -175,3 +199,143 @@ func (m *PostgreSQLBackend) List(prefix string) ([]string, error) {
+			m.key,
+			m.value,
+			m.vaultID,
+		)
```
Rely on the database driver to perform this logic now that lock expiration, `lockTTL`, is of type `time.Duration`.

```go
_, err := m.client.ExecContext(ctxt, lockSQL, m.key, m.value, m.vaultID, m.lockTTL.String())
```

louis-paul commented on physical/postgresql.go:
```diff
@@ -175,3 +199,143 @@ func (m *PostgreSQLBackend) List(prefix string) ([]string, error) {
+		)
+		if err != nil {
+			return nil, err
+		}
```
The locking logic here is to use the table's primary key (on column `key`) as a mean to force mutual exclusion. I think server races should not be an issue (the `DELETE` is done only against already expired locks). With this current logic the `DELETE` has to be performed before the `INSERT`, otherwise the `INSERT` may fail in case a previous lock is still present. Could you expand on how exclusion would be done with `INSERT` before `DELETE`?

Changing the transaction isolation level is something I definitely missed. What implementation would you suggest? For all connections from Vault (with the role used to connect to PostgreSQL or at connection time) or for all individual lock-related transactions? I'm very unfamiliar with the performance concerns here.

sean- replied:
The locking protocol could race if two servers start at the same time.  Also, the `key` isn't very sophisticated so I wouldn't recommend using it for much of anything right now (it's important to make it part of the lock, but I wouldn't rely on it exclusively):

https://github.com/hashicorp/vault/blob/f75f5b0addf00cd0ea16ff0c03f8ae93398ae9e5/vault/core.go#L39-L41
https://github.com/hashicorp/vault/blob/f75f5b0addf00cd0ea16ff0c03f8ae93398ae9e5/vault/core.go#L736-L737
https://github.com/hashicorp/vault/blob/f75f5b0addf00cd0ea16ff0c03f8ae93398ae9e5/vault/core.go#L1445

I think it's necessary to do something like the following:

1. I'd change the table definition to be something like the following ([range types FTW](https://www.postgresql.org/docs/current/static/rangetypes.html)):

```sql
-- Optional:
--
-- CREATE EXTENSION IF NOT EXISTS btree_gist;
--
-- if you want to incorporate the `key` into the lock domain - good practice, not required.  Not all installations have the `btree_gist` extension.

CREATE TABLE vault_lock (
  vault_id   TEXT COLLATE "C" NOT NULL,
  key        TEXT COLLATE "C",
  value      TEXT COLLATE "C",

  -- lock_lease
  lock_lease tstzrange NOT NULL,

  -- row_expiration is the time that we should unconditionally remove this row.  row_expiration should be at least 2x lock_ttl.
  row_expiration TIMESTAMP WITHOUT TIME ZONE NOT NULL DEFAULT NOW() + '30 minutes'::INTERVAL,
  EXCLUDE USING gist (/* key WITH = ,*/ lock_lease WITH &&)
);

CREATE INDEX vault_id_idx ON vault_lock(vault_id);
-- CREATE INDEX vault_lock_row_expiration_idx ON vault_lock(row_expiration); -- only suggested if the row_expiration is very large and a large number of rows accumulate.  This is not necessary but couldn't hurt much.
CREATE INDEX vault_key_idx ON vault_lock(vault_id) WHERE key IS NULL;
```

2. The easy part is is cleaning old rows: `DELETE FROM vault_lock WHERE row_expiration < NOW()`.  Simple and easy to understand.  This could be run by any node (ideally just the one that has the lock _after_ it acquires or renews its lock).

3. Registration of a node is a prerequisite: `INSERT INTO vault_lock (key, value, vault_id) VALUES ($1, $2, $3)`

4. Lock acquisition is an `UPDATE`: `UPDATE vault_lock SET key = $1, lock_lease = ('\'[' || NOW() || ',' || NOW() + $2 || ']\'')::tstzrange WHERE vault_id = $3 AND key IS NULL`;

5. If the above lock has been acquired, you can expire all old rows (query from step 2) `DELETE`.

6. Lock renewal is the following `UPDATE`: `UPDATE vault_lock SET lock_lease = (\'[' || lower(lock_lease) || ',' || NOW() + $1 || ']\'')::tstzrange WHERE vault_id = $2 AND key = $3`

For followers attempting to acquire the lock, poll using:

```sql
SELECT TRUE AS lock_held, vault_id AS lock_held_by, upper(lock_lease) - NOW() AS remaining_lock_duration FROM vault_lock WHERE lock_lease @> NOW()
```

If zero rows come back, attempt to acquire the lock.  If the lock acquisition fails, there was a race and someone else acquired the lock and the follower should go back to polling.  It's tempting to use the `remaining_lock_duration` as the time that the followers should sleep but that also means that the lock won't be acquired until the the lease expires.  Maybe do that if the `poll_interval` is set to `0s`.  ?  Food for thought.

bdarnell replied:
FYI, CockroachDB does not currently support range types. They're not on our roadmap yet, but basic range type support would be fairly straightforward to add (especially after our array support lands). Indexing range types, and the `EXCLUDE USING gist` clause would be harder to support, though.

sean- replied:
@bdarnell What do you think about this:

```sql

CREATE TABLE vault_lock (
  id BIGSERIAL NOT NULL PRIMARY KEY,
  vault_id   TEXT COLLATE "C" NOT NULL,
  key        TEXT COLLATE "C",
  value      TEXT COLLATE "C",

  -- lock_lease
  lock_lease_start TIMESTAMP WITH TIME ZONE,
  lock_lease_end TIMESTAMP WITH TIME ZONE,

  -- row_expiration is the time that we should unconditionally remove this row.  row_expiration should be at least 2x lock_ttl.
  row_expiration TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW() + '30 minutes'::INTERVAL
);

BEGIN;
-- INSERT still happens

-- Perform an update - will crdb's serializable catch this and prevent conflicts?
-- I think this should dirty all of the rows in order to achieve the desired effect.
UPDATE vault_lock SET lock_lease_end = NOW() + $1::INTERVAL
WHERE
  vault_id = $2 AND
  (
    lock_lease_end IS NULL OR lock_lease_end < NOW()) AND
    vault_id NOT IN(
      SELECT vault_id
      FROM vault_lock
      WHERE vault_id != $3 AND (lock_lease_end IS NULL OR lock_lease_end < NOW()
    )
  )
;
COMMIT;
```

bdarnell replied:
Sorry, I'm having a hard time following that transaction. What exactly is the `INSERT` that still happens? Are `$2` and `$3` the same or different? Can you explain what the `NOT IN` clause is trying to do (the double negation of `NOT IN` and `vault_id !=` is confusing me)?

With `SERIALIZABLE` transactions (in both postgres and CRDB), it's fairly simple to do this as a multi-step transaction:

    BEGIN;
    -- SERIALIZABLE is the default on CRDB but not on postgres
    SET TRANSACTION ISOLATION SERIALIZABLE;
    SELECT MAX(lock_lease_end) FROM vault_lock;
    -- if MAX(lock_lease_end) > NOW(), exit
    UPDATE vault_lock SET lock_lease_end = lock_lease_start + $1::INTERVAL WHERE vault_id=$2;
    COMMIT;

The performance of this transaction won't be great (it scans the entire vault_lock table since there is no index, and in both databases it may need to be retried in a loop to ensure serializability), but for occasional lock acquisition that's probably fine. Rewriting it as an UPDATE with a subquery can improve performance significantly (at least in CRDB) if you find that to be worth the complexity. 

Note that if there were an index on `lock_lease_end`, the `SELECT` statement in my transaction would not touch all the rows, just one row plus the gap in the index where any new lease would have to be inserted. In a serializable transaction, that's enough to ensure consistency.

Finally, note that `NOW()` can move backwards (in any database you may experience negative NTP adjustments; in CockroachDB you can additionally find that different nodes have slightly different clocks), so you need to be careful whenever you use `NOW()` that you also watch for times already in the database that may be in your "future".

louis-paul commented:
Sean, thanks for the extensive review. I've fixed most of the style/configuration issues you mentioned. Let me know if there are more.

I have changed the lock watching logic, it would be great to see if that's what you are looking for. Something I have not changed yet is the main locking logic, I've replied to your comment with my questions.

louis-paul commented on physical/postgresql.go:
```diff
@@ -376,17 +382,21 @@ func (m *PostgreSQLLock) watch() {
 			r, err := m.client.Exec(refreshLockSQL, m.lockTTL.String(), m.key,
 				m.vaultID)
 			if err != nil || r == nil {
```
Please have a look at the new version, let me know if that's what you meant.

louis-paul commented on physical/postgresql.go:
```diff
@@ -50,17 +58,68 @@ func newPostgreSQLBackend(conf map[string]string, logger log.Logger) (Backend, e
 	}
 
-	unquoted_table, ok := conf["table"]
+	unquoted_table, ok := conf["lockTableName"]
```
That was some heavy-handed refactoring with my editor 🙂

louis-paul commented:
I revamped the locking logic using your suggestions @sean-. I could not wrap my head around your CockroachDB-compatible transaction, but I tested the SQL statements against CockroachDB v1.0.4. After after changing a bit the table, they seem compatible.

All instances now register in the table when starting the locking process. Active locks are represented with a `lock_lease_end` field that is after `NOW()`. Instances check for the availability of the lock and grab it within the same `SERIALIZABLE` transaction. I removed the `lock_lease_start` which did not seem useful for now. I assumed the `row_expiration` field you made mention of is a mean of garbage collecting dead Vault instances and added logic for that purpose. I’ve hard-coded `row_expiration` to be 2 × `lock_ttl` but we could add a configuration setting for that.

sokoow commented:
Bump, where are we with this - will it ever get merged ?

jefferai commented:
@sokoow waiting on review/discussion.

mytototo commented:
Any news regarding the merge of HA feature for PostgreSQL ?

jefferai commented:
@louis-paul why'd you close it?

louis-paul commented:
@jefferai: There has not been any significant activity for a over a year. Unless I’m mistaken, this is despite the latest review comments having been addressed 😕

jefferai commented:
@louis-paul Probably better to just try pinging @sean- again instead of closing it :-)

jefferai commented:
(P.S. I've also pinged @sean- through a back channel...hopefully I'll at least hear whether or not he is going to review this, and then we can proceed from there)

glerchundi commented:
This would be absolutely fantastic if it can get into Vault, good job @louis-paul and it's also CockroachDB-compatible!

I know that probably everyone is busy from the Hashicorp's side but having a estimated time of action would help making this process a little bit more community-friendly. I also experienced the same feeling some time ago when my PR took 2 years long til got it merged (https://github.com/hashicorp/terraform/pull/3858).

At that time @grubernaut told me that [they were trying to improve](https://github.com/hashicorp/terraform/pull/3858#issuecomment-303109029) the reviewing process but, sadly, this PR shows that nothing has changed since then.

This makes that for example me, as an independent developer, will think twice before starting a contribution process on any Hashicorp's projects.

/cc @mitchellh @armon

jefferai commented:
> I know that probably everyone is busy from the Hashicorp's side but having a estimated time of action would help making this process a little bit more community-friendly. I also experienced the same feeling some time ago when my PR took 2 years long til got it merged (hashicorp/terraform#3858).
> At that time @grubernaut told me that they were trying to improve the reviewing process but, sadly, this PR shows that nothing has changed since then.

Please note that Terraform is a totally separate product with totally separate workflows, and Jake has not been with the company for a long time. I don't know what the holdup was with Terraform, but the problem with this issue is that HashiCorp explicitly does not maintain this storage backend. We do basic due diligence, but we don't test it out, and we aren't PG experts. So we try to get people that actually are using it and/or *are* PG experts, other than the PR submitter, to review it so that merging something doesn't break everybody else.

The problem right now is that the person who did the review and asked for a number of changes then disappeared, and in the interim time, past the initial ping after the changes were made, nobody tried pinging that person to see if they could come back and take another look.

The person has been pinged here and I've pinged them through a backchannel, but that was on Thursday and it's the first day back from a holiday weekend here in the U.S. It's not unreasonable that there's been no reply yet.

Another possibility would be to find other users of the storage backend (there are many; you could for instance post on the mailing list) and ask some of them to review it. Then you're not bottlenecked on the single reviewer from before.

jefferai commented:
I should also mention that there are merge conflicts which have to be solved before we can merge post-review anyways...

glerchundi commented:
@jefferai now is crystal clear, thanks for giving us an update.

@louis-paul in case you have no time to invest on this I can take it and made the requested changes.

glerchundi commented on physical/postgresql.go:
```diff
@@ -74,6 +156,11 @@ func newPostgreSQLBackend(conf map[string]string, logger log.Logger) (Backend, e
 			"UNION SELECT DISTINCT substring(substr(path, length($1)+1) from '^.*?/') FROM " +
 			quoted_table + " WHERE parent_path LIKE concat($1, '%')",
 		logger: logger,
+
```
pinging @knz which is one of the master & commanders of CockroachDB in case he says this is not supported yet and if he has an idea of how could be rewritten in a way which it is.

knz replied:
I'm not exactly sure what the question is? You already use `pq.QuoteIdentifier` properly; you could format your query with `fmt.Sprintf("SELECT value FROM %s WHERE path = $1 AND key = $2", quoted_table)` and that should work. Was there any reason to not do that?

Meanwhile regarding the `list_query`:

- the UNION should work; however
- I'm not sure what you are doing with `substring( ... from '^.*?/')`. This does't look right; `substring` does not work with regular expressions. There are `regexp_xxx()` functions that can help though.

glerchundi replied:
It was for the `UNION SELECT DISTINCT` thingy so thanks for the quick response!

jefferai commented:
@glerchundi It seems like the OP isn't interested -- if you can address this feedback it'd be great!

glerchundi commented:
I’m on vacation now but will take this and address the issues once I find a
free slot. This means in a month or so.

Thanks for pinging.

On Thu, 4 Oct 2018 at 16:58, Jeff Mitchell <notifications@github.com> wrote:

> @glerchundi <https://github.com/glerchundi> It seems like the OP isn't
> interested -- if you can address this feedback it'd be great!
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/hashicorp/vault/pull/2700#issuecomment-427051564>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACIPlsYF6OTFfM2vveogqB-LbtSJnONtks5uhiIcgaJpZM4NVlhF>
> .
>

bjorndolk commented:
Hello!,  I started working on addressing feedback in this PR. The filestructure for under physical/... has been refactored since this PR was opened. It is kind of a mess to merge.

bjorndolk commented:
We (silverrail) may also work on a docker image for postgres to get this more testable, but we will see about that.

louis-paul commented:
Hi @bjorndolk, I’ve added you as a collaborator to the fork’s repo (closing this PR and opening a new one is also fine).

I have looked at updating the branch, but the amount of work is non-trivial and I found my code’s quality to be worse than I expected. Sorry for the inactivity on this change, but I can’t justify spending more time on this as my company has dropped Vault.

bjorndolk commented:
@louis-paul Thanks !!  I got this merged and compiling now. Next is to add the generic HA tests and see how that goes.

bjorndolk commented:
Running the genric tests, normal backend tests still work, so didnt break that in merge. 
However the HA tests breaks, these tests did not exist when this was originally written. I will dig into this.

bjorndolk commented:
@jefferai  please consider #5731 for postgres ha support.

jefferai commented:
Closed this as the only person that seems interested in getting it across the finish line has rewritten it in a new PR (linked above).

glerchundi commented:
> Sorry for the inactivity on this change, but I can’t justify spending more time on this as my company has dropped Vault.

Could I ask what are you using instead of Vault, hand rolled or another soft? We're evaluating the use of Vault and would be good to know why people drop Vault!

Thanks!
